value/cast.ts

```diff
@@ -0,0 +1,416 @@
+function cast(value: Value, from_: Type.Real, to_: Type.String): string;
+function cast(
+  value: Value,
+  from_: Type.Real,
+  to_: Type.Exception,
+): BaseException;
+function cast(value: Value, from_: Type.Boolean, to_: Type.Integer): number;
+function cast(value: Value, from_: Type.Boolean, to_: Type.Real): number;
+function cast(value: Value, from_: Type.Boolean, to_: Type.Boolean): boolean;
+function cast(value: Value, from_: Type.Boolean, to_: Type.String): string;
+function cast(
+  value: Value,
+  from_: Type.Boolean,
+  to_: Type.Exception,
+): BaseException;
+function cast(value: Value, from_: Type.String, to_: Type.Integer): number;
+function cast(value: Value, from_: Type.String, to_: Type.Real): number;
+function cast(value: Value, from_: Type.String, to_: Type.Boolean): boolean;
+function cast(value: Value, from_: Type.String, to_: Type.String): string;
+function cast(
+  value: Value,
+  from_: Type.String,
+  to_: Type.Exception,
+): BaseException;
+function cast(value: Value, from_: Type.Exception, to_: Type.Integer): number;
+function cast(value: Value, from_: Type.Exception, to_: Type.Real): number;
+function cast(value: Value, from_: Type.Exception, to_: Type.Boolean): boolean;
+function cast(value: Value, from_: Type.Exception, to_: Type.String): string;
+function cast(
+  value: Value,
+  from_: Type.Exception,
+  to_: Type.Exception,
+): BaseException;
+function cast(value: Value, from_: Type.Nil, to_: Type.Integer): number;
+function cast(value: Value, from_: Type.Nil, to_: Type.Real): number;
+function cast(value: Value, from_: Type.Nil, to_: Type.Boolean): boolean;
+function cast(value: Value, from_: Type.Nil, to_: Type.String): string;
+function cast(
+  value: Value,
+  from_: Type.Nil,
+  to_: Type.Exception,
+): BaseException;
+function cast(value: Value, from_: Type.Any, to_: Type.Integer): number;
+function cast(value: Value, from_: Type.Any, to_: Type.Real): number;
+function cast(value: Value, from_: Type.Any, to_: Type.Boolean): boolean;
+function cast(value: Value, from_: Type.Any, to_: Type.String): string;
+function cast(
+  value: Value,
+  from_: Type.Any,
+  to_: Type.Exception,
+): BaseException;
+function cast(value: Value, from_: Type, to_: Type): never;
+function cast(value: Value, from_: Type, to_: Type): Value {
+  switch (from_) {
+    case Type.Integer:
+      return castInteger(value as number, to_);
+    case Type.Real:
+      return castReal(value as number, to_);
+    case Type.Boolean:
+      return castBoolean(value as boolean, to_);
+    case Type.String:
+      return castString(value as string, to_);
+    case Type.Exception:
+      return castException(value as BaseException, to_);
+    case Type.Nil:
+      return castNil(value as Nil, to_);
+    case Type.Any:
+      return castAny(value as Value, to_);
+    default:
+      return castFault(value, UNKNOWN, UNKNOWN);
+  }
+}
+
+function wouldCast(from_: Type, to_: Type): boolean {
```

Comment:
There are actually three different potential results here - the boolean accounts for "yes it would" and "no it would throw", but right now "we don't know right now" is signalled by a thrown error.

But I probably don't want to throw an error when the result is that it *would* fail. The choice is basically between:

```ts
let would: boolean | null = null;
try {
 would = wouldConvert(a, b);
} catch (err) {
  if (err instanceof TypeError) {
    // Know that it *might* convert at runtime
  }
}

if (would !== null) {
  if (would) {
    // Know that it would successfully convert
  } else {
    // Know that it would fail to convert
  }
}
```

and:

```ts
const would = wouldConvert(a, b);

if (would === 'yes') {
  // Know that it would successfully convert
} else if (would === 'no') {
  // Know that it would fail to convert
} else if (would === 'maybe') {
  // Know that it might convert at runtime
}
```